it/src/test/java/com/google/cloud/teleport/it/spanner/SpannerResourceManagerUtilsTest.java

```diff
@@ -0,0 +1,195 @@
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseIdShouldTrimTrailingHyphen() {
+    String testBaseString = "test_database---";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseIdShouldTrimTrailingUnderscore() {
+    String testBaseString = "test_database___";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseIdShouldThrowErrorWithEmptyInput() {
+    String testBaseString = "";
+
+    assertThrows(IllegalArgumentException.class, () -> generateDatabaseId(testBaseString));
+  }
+
+  @Test
+  public void testGenerateDatabaseIdShouldThrowErrorWhenInputContainsNoLettersOrNumbers() {
+    String testBaseString = "---___$...__";
+
+    assertThrows(IllegalArgumentException.class, () -> generateDatabaseId(testBaseString));
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldReplaceDollarSignWithHyphen() {
+    String testBaseString = "test$instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-\\d{8}-\\d{6}-\\d{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldReplaceDotWithHyphen() {
+    String testBaseString = "test.instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-\\d{8}-\\d{6}-\\d{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldReplaceNonLetterFirstCharWithLetter() {
+    String testBaseString = "0-test-instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("[a-z]-test-instance-\\d{8}-\\d{6}-\\d{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldReplaceUnderscoreWithHyphen() {
+    String testBaseString = "test_instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-\\d{8}-\\d{6}-\\d{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldReplaceUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test-Instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-\\d{8}-\\d{6}-\\d{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceIdShouldThrowErrorWithEmptyInput() {
+    String testBaseString = "";
+
+    assertThrows(IllegalArgumentException.class, () -> generateInstanceId(testBaseString));
+  }
+
+  @Test
+  public void testGenerateNewIdShouldReturnNewIdWhenInputLengthIsLongerThanTargetLength() {
+    String longId = "long-test-id-string";
+
+    String actual = generateNewId(longId, 13);
+
+    assertThat(actual.length()).isAtMost(13);
+    assertThat(actual).matches("long-([a-zA-Z]|\\d)*");
```

Comment:
This could be simplified a bit: `"long-[a-zA-Z\\d]{8}"` (This assumes the length is always 13.)

You can also do `"long-[a-zA-Z0-9]{8}"` if you'd rather stay consistent with using ranges, since the `\\d` shorthand breaks from that.